src/Interpreters/SessionTracker.cpp

```diff
@@ -0,0 +1,62 @@
+#include "SessionTracker.h"
+
+#include <Common/Exception.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int USER_SESSION_LIMIT_EXCEEDED;
+}
+
+SessionTracker::Session::Session(SessionTracker & tracker_,
+                                 const UUID& user_id_,
+                                 SessionInfos::const_iterator session_info_iter_) noexcept
+    : tracker(tracker_), user_id(user_id_), session_info_iter(session_info_iter_)
+{
+}
+
+SessionTracker::Session::~Session()
+{
+    tracker.stopTracking(user_id, session_info_iter);
+}
+
+SessionTracker::SessionTrackerHandle
+SessionTracker::trackSession(const UUID & user_id,
+                             const SessionInfo & session_info,
+                             size_t max_sessions_for_user)
+{
+    std::lock_guard lock(mutex);
+
+    auto sessions_for_user_iter = sessions_for_user.find(user_id);
+    if (sessions_for_user_iter == sessions_for_user.end())
+        sessions_for_user_iter  = sessions_for_user.emplace(user_id, SessionInfos()).first;
+
+    SessionInfos & session_infos = sessions_for_user_iter ->second;
+    if (max_sessions_for_user && session_infos.size() >= max_sessions_for_user)
+    {
+        throw Exception(ErrorCodes::USER_SESSION_LIMIT_EXCEEDED,
+                        "User {} has overflown session count {}",
+                        session_info.client_info.current_user,
+                        max_sessions_for_user);
+    }
+
```

Comment:
Since you added an assert into `stopTracking()`, does it make sense to add another assert here that the input `session_info` is not already contained in `session_infos`?

Reply:
I think we don't need such assert.
in stopTracking() I check that `user_id `exist. 
In case of `trackSession` It is possible that **user_id** exist and not exist. 

I don't check that `session_id `is exist in a list, because this operation is `O(n)`